### PR TITLE
sayAll: no longer prematurely stop sayAll after turning a page.

### LINKED ISSUE #14390: NVDA only reads one page at a time in Say all in Kindle App
### Actual behavior:

When I try to read a Kindle book using Say All mode, NVDA only reads one page of a book and then stops talking.

When I shift the focus to the start of the nexet page and start Say All again, NVDA will read until it gets to the end of the page and then it stops talking again.

Here is a link to a book you can try.

https://www.amazon.com/Great-Britain-American-Civil-War-ebook/dp/B0082VDZY2/ref=sr_1_3?keywords=Great+Britain+and+the+American+Civil+War&qid=1669126627&s=digital-text&sr=1-3

### Expected behavior:

Say all should read to the end of the book if not stopped.

#### NVDA version:
 NVDA V2022.4 beta 3

#### Windows version:
Windows 11 latest insider build
#### Name and version of other software in use when reproducing the issue:
Kindle

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #14390

### Summary of the issue:
With the merging of pr #14070 which added sayall in tables, SayAll in Kindle for PC would fail to continue reading after turning the page.
Technical:
Some of the code in the nextLine method in sayAll was refacted into a nextLineImpl method. 
However, if nextLineImpl returned false, finish would be called and nextLIne would return. This would be correct for handling the case where there was no more text, or table cells, but not the case where the page has just been turned. In fact, text sayAll already called finish when there was no more text, so calling finish again was useless in the base case, but for page turns caused sayAll to abort prematurely after the page was turned.

### Description of user facing changes
NVDA no longer fails to keep reading with sayAll after crossing a page boundary.

### Description of development approach
In sayAll's nextLine method, removed the call to self.finish when nextLineImpl returns False, but ensured that  table sayAll's nextLineImpl does call finish itself if there is no more table cells.
In other words, the self.finish call has been moved into the specific nextLineImpl method where it is needed, rather than running more broadly.

### Testing strategy:
Opened a book in Kindle for PC. Started a sayAll and ensured that NVDA continued reading after the page was turned.
Using the table on the NvDA   snapshots page, ensured that NvDA read all the way across the first row using NVDA+control+alt+rightArrow, and that sayAll profile was handled correctly I.e. the synth was reset to the normal voice after sayAll completed.

### Known issues with pull request:
None.

### Change log entries:
New features
Changes
Bug fixes
- NVDA no longer fails to continue reading in Kindle for PC after turning the page. (#14390) 

For Developers

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by Brian1Gaff** [comment]
Will this appear in an rc2 of the current rc cycle, or will it need a new 
version?
 Brian

-- 
***@***.***
Sent via blueyonder.(Virgin media)
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Michael Curran" ***@***.***>
To: "nvaccess/nvda" ***@***.***>
Cc: "Subscribed" ***@***.***>
Sent: Friday, March 10, 2023 6:03 AM
Subject: [nvaccess/nvda] sayAll: no longer prematurely stop sayAll after 
turning a page. (PR #14711)

**Comment by michaelDCurran** [comment]
I'm hoping an rc2.

**Comment by Brian1Gaff** [comment]
Good, I had no idea there were so many Avid Kindle users until now. I find I 
don't have time to read audio books these days.
 Brian

-- 
***@***.***
Sent via blueyonder.(Virgin media)
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Michael Curran" ***@***.***>
To: "nvaccess/nvda" ***@***.***>
Cc: "Brian Gaff" ***@***.***>; "Comment" 
***@***.***>
Sent: Saturday, March 11, 2023 5:00 AM
Subject: Re: [nvaccess/nvda] sayAll: no longer prematurely stop sayAll after 
turning a page. (PR #14711)

**Review by seanbudd — APPROVED** [review]
